proxy/iptables: don't sync proxy rules if services map didn't change

dcbw opened the pull request:
Build the service map in a separate testable function.  Return that map instead of changing proxier.serviceMap directly.  Use reflect.DeepEqual() to skip syncing proxy rules if nothing actually changed.

@thockin @kubernetes/rh-networking @kubernetes/sig-network-misc @timothysc @wojtek-t @jeremyeder @caseydavenport

dcbw committed:
proxy/iptables: split out service map creation and add testcases

dcbw committed:
proxy/iptables: don't proxy ExternalName services

The API docs say:

	// ServiceTypeExternalName means a service consists of only a reference to
	// an external name that kubedns or equivalent will return as a CNAME
	// record, with no exposing or proxying of any pods involved.

which implies that ExternalName services should be ignored for proxy
purposes.

dcbw committed:
proxy/iptables: don't sync proxy rules if services map didn't change

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/38996)
<!-- Reviewable:end -->

ravilr commented:
@dcbw this fixes https://github.com/kubernetes/kubernetes/issues/26637 ?

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -423,16 +425,24 @@ func (proxier *Proxier) SyncLoop() {
+	namespace types.NamespacedName
+	nodeport  int
+	// if true, should be added to healthcheck listener service.  If false,
+	// shoul dbe deleted.
```
typo
(also, kind of an awkward API. Maybe return separate add/remove arrays rather than a single array of adds-and-removes?)

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -457,15 +467,15 @@ func (proxier *Proxier) OnServiceUpdate(allServices []api.Service) {
-				// Nothing changed.
-				continue
-			}
+			info, exists := newServiceMap[serviceName]
```
It's weird to check newServiceMap to see if the service existed in oldServiceMap. I think the loop would make more sense if you started with an empty newServiceMap and copied pre-existing-unchanged services over from oldServiceMap, rather than starting with a complete copy of oldServiceMap and then deleting everything that had changed.

dcbw replied:
FWIW latest pushes address this in isolated git commits.

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -454,10 +454,15 @@ func buildServiceMap(allServices []api.Service, oldServiceMap proxyServiceMap) (
 			continue
 		}
+		// Even if ClusterIP is set, ServiceTypeExternalName services don't get proxied
+		if service.Spec.Type == api.ServiceTypeExternalName {
```
you're testing the same case twice here?

dcbw replied:
Yeah, meant to remove it from the if above.

danwinship reviewed (commented):
nice. much simpler than the old version...

dcbw commented:
> nice. much simpler than the old version...

@danwinship this is in addition to the old version :)  But it's possible that with this, and with an endpoints update PR that I'm still working on, that the cost/benefit ratio of the iptables-parsing one isn't great enough to justify a merge...

dcbw commented:
> @dcbw this fixes #26637 ?

@ravilr it's one component of a fix; this only does services, I'm going to do another PR for the endpoints too.  That possibly combined with https://github.com/kubernetes/kubernetes/pull/36006 should be a more complete solution.

danwinship commented:
> @danwinship this is in addition to the "old version" :) But it's possible that with this, and with an endpoints update PR that I'm still working on, that the cost/benefit ratio of the iptables-parsing one isn't great enough to justify a merge...

Oh, right, Services not Endpoints. But you ought to be able to do the same thing for the endpoints: if the relevant subset of the endpoint data is the same as it was before, then you know the set of generated iptables rules would be the same too, so there's no need to compare against the *actual* iptables rules

dcbw commented:
> Oh, right, Services not Endpoints. But you ought to be able to do the same thing for the endpoints: if the relevant subset of the endpoint data is the same as it was before, then you know the set of generated iptables rules would be the same too, so there's no need to compare against the actual iptables rules

@danwinship yes, this is already in-progress :)

I actually started with the services/endpoints checking thing.  The reason I then switched to comparing against the actual iptables rules was to ensure that the rules didn't get changed out underneath kubelet, as we've seen time-and-again with OpenShift.  Not checking iptables rules directly (like SyncProxyRules() currently does) would lead to a large lag in correcting the modified iptables rules leaving a broken cluster for potentially minutes or longer.

I spoke to @thockin about that, and he convinced me that's not a huge deal because the iptables code will periodically force-sync the rules anyway. Though as we've seen that force-sync might only happen every 5 minutes or longer depending on the --iptables-sync-period flag.  But TIm doesn't think that's a huge problem, so here we are.

dcbw committed:
proxy/iptables: clean up service map creation

Instead of copying the map, like OnServicesUpdate() used to do and which
was copied into buildServiceMap() to preserve semantics while creating
testcases, start with a new empty map and do deletion checking later.

dcbw commented:
@danwinship PTAL, thanks!

dcbw commented:
Jenkins says: "Unable to load build details from ..." and gives no info.

@k8s-bot unit test this

dcbw commented:
@danwinship @kubernetes/sig-network-misc @thockin @timothysc @jeremyeder PTAL; latest push adds endpoint map sync reduction too.

dcbw commented:
@k8s-bot cvm gke e2e test this issue https://github.com/kubernetes/kubernetes/issues/35913

dcbw commented:
@k8s-bot kubemark e2e test this

dcbw commented:
@k8s-bot gci gce e2e test this

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -486,42 +494,64 @@ func (proxier *Proxier) OnServiceUpdate(allServices []api.Service) {
+	for _, hc := range hcAdd {
+		glog.V(4).Infof("Adding health check for %+v, port %v", hc.namespace, hc.nodeport)
+		// Turn on healthcheck responder to listen on the health check nodePort
+		healthcheck.AddServiceListener(hc.namespace, hc.nodeport)
```
This highligts that this could maybe fail without error or retry...

dcbw replied:
@thockin added a FIXME so we don't forget it

thockin reviewed (commented):
This looks OK.  I like @danwinship comments, too.  This could be a PR of it's own.  Get it approved in isolation, get it off this PR

thockin commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1121,4 +1121,44 @@ func TestBuildServiceMapServiceHeadless(t *testing.T) {
+func TestBuildServiceMapServiceTypeExternalName(t *testing.T) {
+	services := []api.Service{
+		{
+			ObjectMeta: api.ObjectMeta{
```
For this test and the other's like it, it might be really nice to avoid boilerplate clutter and do something like:

```
services := []api.Service{
    makeService("external-name", func(svc *api.Service) {
        svc.Spec.Type = api.ServiceTypeExternalName
        svc.Spec.ExternalName = "foo2.bar.com"
        svc.Spec.ClusterIP = "172.16.55.4"
    })
```

That way the boilerplate struct gunk is the same (or randomized) and the testcase ONLY carries the things that are unique, which makes it easier to comprehend each case.

dcbw replied:
Done.

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -555,7 +553,11 @@ func (proxier *Proxier) OnServiceUpdate(allServices []api.Service) {
 	}
 
-	proxier.syncProxyRules()
+	if !reflect.DeepEqual(newServiceMap, proxier.serviceMap) {
```
might be an optimization to compare the len() first and if not equal, skip the reflection, or even to compare keys.

dcbw replied:
Done.

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -555,7 +553,11 @@ func (proxier *Proxier) OnServiceUpdate(allServices []api.Service) {
+	if !reflect.DeepEqual(newServiceMap, proxier.serviceMap) {
+		proxier.serviceMap = newServiceMap
+		proxier.syncProxyRules()
+	}
```
log if this is a skip, maybe V(3) ?

dcbw replied:
Done.

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -500,26 +456,38 @@ func buildServiceMap(allServices []api.Service, oldServiceMap proxyServiceMap) (
+			}
+
+			oldInfo, exists := oldServiceMap[serviceName]
+			equal := reflect.DeepEqual(info, oldInfo)
```
only do the DeepEqual if REALLY necessary, i.e. after proving `exists == true`

dcbw replied:
I checked that before doing this, I don't think it's necessary as DeepEqual already checks for nil and exits early.  Basically doing exactly what I would do if I open-coded it here.

```
func DeepEqual(x, y interface{}) bool {

		if x == nil || y == nil {

			return x == y

		}
```

thockin reviewed (commented):
This commit feels like maybe it could merge with #1, bu maybe subtle conflicts make it hard?

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -564,37 +569,34 @@ func (proxier *Proxier) buildEndpointInfoList(endPoints []hostPortInfo, endpoint
-				port := &ss.Ports[i]
-				for i := range ss.Addresses {
-					addr := &ss.Addresses[i]
+		for _, ss := range svcEndpoints.Subsets {
```
I did the loops with pointers before because Addresses in particular can get large

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -657,12 +657,15 @@ func (proxier *Proxier) OnEndpointsUpdate(allEndpoints []api.Endpoints) {
 	}
 
-	proxier.syncProxyRules()
+	if !reflect.DeepEqual(endpointsMap, proxier.endpointsMap) {
```
same optimization comments - avoid reflect if we can, so pre-filter

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -569,15 +569,26 @@ func buildEndpointInfoList(endPoints []hostPortInfo, endpointIPs []string) []*en
-	endpoints []hostPortInfo
+// Map of namespaced names (eg, service namespace/name) to whether the service
+// has local endpoints or not.  Only local endpoints need healthchecking.
+type hcUpdateMap map[types.NamespacedName]sets.String
```
Comment is not an accurate description of the type. "Map from a namespaced name to either an empty set or a set containing the same namespaced name as a string, because that's the type we're going to need later, because healthcheck.UpdateEndpoints() is a silly API"...

dcbw replied:
Done

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -585,6 +564,22 @@ func addHcUpdate(updateMap hcUpdateMap, name types.NamespacedName, isLocal bool)
+func addEndpointToMap(endpointsMap map[string]endpointsInfoSlice, ip string, port api.EndpointPort, isLocal bool) {
+	// Only add the hostport if the address is valid,
+	// but keep track of port names with invalid addresses
+	// so we can clean them up later
```
More lies! You just log invalid endpoints and then forget about them.

dcbw replied:
Actually no, that was the point of logging but then assigning the empty endpointsInfoSlice to the map:

`endpointsMap[port.Name] = endpoints`

even if `ip` or `port` are invalid.  We do need it later.

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -611,30 +606,22 @@ func buildEndpointsMap(allEndpoints []api.Endpoints, hostname string, oldEndpoin
-			if len(curEndpointIPs) != len(newEndpoints) || !slicesEquiv(slice.CopyStrings(curEndpointIPs), newEndpoints) {
-				removedEndpoints := getRemovedEndpoints(curEndpointIPs, newEndpoints)
+			sort.Sort(newEndpoints)
+			if !reflect.DeepEqual(curEndpoints, newEndpoints) || len(newEndpoints) == 0 {
```
The `len(newEndpoints) == 0` check seems unnecessary

dcbw replied:
@danwinship you're right, it can be removed if we move the `newEndpointsMap[svcPort] = newEndpoints` out of the if{} block.

thockin commented:
ping me when this is ready again

dcbw commented:
> This commit feels like maybe it could merge with #1, bu maybe subtle conflicts make it hard?

@thockin I'd rather keep them separate as they address separate independent problems.  In fact I'm going to split that out to a second PR per your request before the holidays.

dcbw commented:
@thockin @danwinship repushed and updated for Tim's comments.  I have dropped the Endpoint update series as I'll submit that as a second PR per Tim's request, and when doing so I'll address the comments from this PR in that one.

dcbw commented:
@k8s-bot kops aws e2e test this

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -153,18 +153,28 @@ type endpointsInfo struct {
+		loadBalancerStatus:       *api.LoadBalancerStatusDeepCopy(&status.LoadBalancer),
+		sessionAffinityType:      spec.SessionAffinity,
+		stickyMaxAgeMinutes:      180, // TODO: paramaterize this in the API.
+		externalIPs:              spec.ExternalIPs,
```
should also deep-copy?

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -153,18 +153,28 @@ type endpointsInfo struct {
+		sessionAffinityType:      spec.SessionAffinity,
+		stickyMaxAgeMinutes:      180, // TODO: paramaterize this in the API.
+		externalIPs:              spec.ExternalIPs,
+		loadBalancerSourceRanges: spec.LoadBalancerSourceRanges,
```
and this?

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -471,21 +448,9 @@ func (proxier *Proxier) OnServiceUpdate(allServices []api.Service) {
-			info.clusterIP = serviceIP
+
+			info := newServiceInfo(serviceName, servicePort, &service.Spec, &service.Status)
+			info.clusterIP = net.ParseIP(service.Spec.ClusterIP)
```
IP, port, etc are all being set in `newServiceInfo()` now, pick one.

thockin replied:
I'm inclined to leave it here, since you also have the only-local stuf here.

dcbw replied:
moved the nodelocal stuff to the function instead.

thockin reviewed (commented):
just a few minor points, then LGTM.  Thanks.  I am marking this for relnote - I think it is significant.

danwinship commented:
lgtm too

dcbw commented:
@k8s-bot unit test this issue #39510

dcbw commented:
@thockin fixed up for your comments.

thockin commented:
Thanks!
